mlflow/server/handlers.py

```diff
@@ -1103,6 +1104,117 @@ def _default_history_bulk_impl():
+    else:
+        max_results = MAX_RESULTS_PER_RUN
+
+    metric_key = request.args.get("metric_key")
+    if metric_key is None:
+        raise MlflowException.invalid_parameter_value(
+            "GetMetricHistoryBulkInterval request must specify a metric_key."
+        )
+
+    store = _get_tracking_store()
+
+    def _get_max_step_for_metric(run_id, metric_key):
+        if hasattr(store, "get_max_step_for_metric"):
+            return store.get_max_step_for_metric(run_id=run_id, metric_key=metric_key)
+        return max(m.step for m in store.get_metric_history(run_id, metric_key))
+
+    def _get_sampled_steps(run_ids, metric_key, max_results):
+        start_step = args.get("start_step")
+        end_step = args.get("end_step")
+        if start_step is None and end_step is None:
+            max_steps_per_run = [_get_max_step_for_metric(run_id, metric_key) for run_id in run_ids]
+            max_steps = max(max_steps_per_run)
+            # Get sampled steps and append the max step for each run.
+            sampled_steps = _get_sampled_steps_from_steps(0, max_steps, max_results)
+            return list(set(sampled_steps).union(max_steps_per_run))
+        elif start_step is not None and end_step is not None:
+            start_step = int(start_step)
+            end_step = int(end_step)
+            if start_step > end_step:
+                raise MlflowException.invalid_parameter_value(
+                    "end_step must be greater than start_step. "
+                    f"Found start_step={start_step} and end_step={end_step}."
+                )
+            sampled_steps = _get_sampled_steps_from_steps(start_step, end_step, max_results)
+            return list(set(sampled_steps).union([end_step]))
+        else:
+            raise MlflowException.invalid_parameter_value(
+                "If either start step or end step are specified, both must be specified."
+            )
+
+    def _default_history_bulk_interval_impl():
+        steps = _get_sampled_steps(run_ids, metric_key, max_results)
+        metrics_with_run_ids = []
+        for run_id in run_ids:
+            metrics_with_run_ids.extend(
+                store.get_metric_history_bulk_interval_from_steps(
+                    run_id=run_id,
+                    metric_key=metric_key,
+                    steps=steps,
+                    max_results=MAX_RESULTS_GET_METRIC_HISTORY,
+                )
+            )
+        return metrics_with_run_ids
+
+    if hasattr(store, "get_metric_history_bulk_interval"):
+        metrics_with_run_ids = [
+            metric.to_dict()
+            for metric in store.get_metric_history_bulk_interval(
+                run_ids=run_ids,
+                metric_key=metric_key,
+                start_step=args.get("start_step"),
+                end_step=args.get("end_step"),
+                max_results=max_results,
+            )
+        ]
+    else:
+        metrics_with_run_ids = _default_history_bulk_interval_impl()
```

Comment:
Do we need this if condition? All stores implement `get_metric_history_bulk_interval_from_steps`, right?

Reply:
This is checking if `get_metric_history_bulk_interval` function exists. But I'm happy to remove it and always use our implementation, so users could only optimize `get_max_step_for_metric` and `get_metric_history_bulk_interval_from_steps` functions.